File: sorting/MergeSort.java
package sorting;

import java.util.ArrayList;
import java.util.List;

public class MergeSort {
    public static int[] sortArray(int[] nums) {
        mergeSort(nums, 0, nums.length - 1);
        return nums;
    }

    public static void merge(int[] nums, int low, int mid, int high) {
        int left = low;
        int right = mid + 1;
        List<Integer> ans = new ArrayList<>();
        while (left <= mid && right <= high) {
            if (nums[left] <= nums[right]) {
                ans.add(nums[left]);
                left++;
            } else {
                ans.add(nums[right]);
                right++;
            }
        }
        while (left <= mid) {
            ans.add(nums[left]);
            left++;
        }
        while (right <= high) {
            ans.add(nums[right]);
            right++;
        }
        // move all elements to the original array
        for (int i = low; i <= high; i++) {
            nums[i] = ans.get(i - low);
        }
    }

    private static void mergeSort(int[] nums, int low, int high) {
        if (low >= high) {
            return;
        }
        int mid = (low + high) / 2;
        mergeSort(nums, low, mid);
        mergeSort(nums, mid + 1, high);
        merge(nums, low, mid, high);
    }

    public static void main(String[] args) {
        int[] nums = {5, 2, 3, 1};
        sortArray(nums);
        for (int i : nums) {
            System.out.println(i);
        }
    }
}

File: tufplus/sorting/MergeSort.py
class Solution:
    def mergeSort(self, nums):
        self.mergeSortAlgo(nums, 0, len(nums) - 1)
        return nums
    
    def mergeSortAlgo(self, nums, low, high):
        # base case
        if low >= high:
            return
        # find mid
        mid = (low + high) // 2
        # divide
        self.mergeSortAlgo(nums, low, mid)
        self.mergeSortAlgo(nums, mid + 1, high)
        # merge
        self.merge(nums, low, mid, high)
    
    def merge(self, nums, low, mid, high):
        # create temp array
        temp = []
        # two pointers
        i=low
        j=mid+1
        # traverse both arrays
        while i<=mid and j<=high:
            if nums[i] <= nums[j]:
                temp.append(nums[i])
                i+=1
            else:
                temp.append(nums[j])
                j+=1
        # copy remaining elements
        while i<=mid:
            temp.append(nums[i])
            i+=1
        while j<=high:
            temp.append(nums[j])
            j+=1
        # copy back to original array

        # for example, if nums=[3,5,7,2,1,4], low=0, high=5, temp=[1,2,3,4,5,7]
        # then, nums will be [1,2,3,4,5,7]
        for i in range(low, high + 1):
            nums[i] = temp[i - low]
        
